core/server/master/src/main/java/alluxio/master/file/LostFileDetector.java

```diff
@@ -37,40 +43,68 @@ final class LostFileDetector implements HeartbeatExecutor {
   private static final Logger LOG = LoggerFactory.getLogger(LostFileDetector.class);
   private final FileSystemMaster mFileSystemMaster;
   private final InodeTree mInodeTree;
+  private final BlockMaster mBlockMaster;
 
   /**
    * Constructs a new {@link LostFileDetector}.
    */
-  public LostFileDetector(FileSystemMaster fileSystemMaster, InodeTree inodeTree) {
+  public LostFileDetector(FileSystemMaster fileSystemMaster, BlockMaster blockMaster,
+      InodeTree inodeTree) {
     mFileSystemMaster = fileSystemMaster;
     mInodeTree = inodeTree;
+    mBlockMaster = blockMaster;
     MetricsSystem.registerCachedGaugeIfAbsent(MetricKey.MASTER_LOST_FILE_COUNT.getName(),
         () -> mFileSystemMaster.getLostFiles().size());
   }
 
   @Override
   public void heartbeat() throws InterruptedException {
-    for (long fileId : mFileSystemMaster.getLostFiles()) {
-      // Throw if interrupted.
+    Iterator<Long> iter = mBlockMaster.getLostBlocksIterator();
+    Set<Long> lostFiles = new HashSet<>();
+    while (iter.hasNext()) {
+      long blockId = iter.next();
+      long containerId = BlockId.getContainerId(blockId);
+      long fileId = IdUtils.createFileId(containerId);
+      if (lostFiles.contains(fileId)) {
+        iter.remove();
+        continue;
+      }
       if (Thread.interrupted()) {
         throw new InterruptedException("LostFileDetector interrupted.");
       }
```

Comment:
nit: Pls move this interrupted check to the beginning of the for loop
```
while (...) {
  if (Thread.interrupted()) {...}
  // continue your workflow
}
```

Reply:
done